### ISSUE TITLE
Title w/separator (eX) rule behaves differently across the different rule engines

### ISSUE THREAD

**Issue by hops** [hashcat/hashcat#1969]
Given the word `foobar` and the rule `ef` I would expect the mangled word to become `FOobar`
You could also argue that since `f` will become `F` as first operation, that the separator isn't present anymore. Therefore I would also accept `Foobar`.
However it is behaving differently in the pure kernel's rule engine, the optimized kernel's rule engine and the one on the host (invoked with -j/-k).

Prepping steps for reproducing:
```
$ echo -n FOobar | md5sum
6eb3fa350652503b81bdfe19ca70d224  -
$ cat rule
ef
```
Version 5.1.0 results:
```
$ ./hashcat --version
v5.1.0
$ echo foobar | ./hashcat --quiet --potfile-disable -r rule 6eb3fa350652503b81bdfe19ca70d224
Status...........: Exhausted

$ echo foobar | ./hashcat --quiet --potfile-disable -O -r rule 6eb3fa350652503b81bdfe19ca70d224
6eb3fa350652503b81bdfe19ca70d224:Foobar
Status...........: Cracked

$ echo foobar | ./hashcat -r rule --stdout
Foobar
$ echo foobar | ./hashcat -O -r rule --stdout
Foobar
$ echo foobar | ./hashcat -j ef --stdout
fOobar
```
To recap: The pure kernel doesn't crack the hash. The optimized kernel does but reports the wrong password. And -j is doing whatever?
@Chick3nman did get the same results with v 4.2.1

Now the current master:
```
$ ./hashcat --version
v5.1.0-795-g7806be32
$ echo foobar | ./hashcat --quiet --potfile-disable -r rule 6eb3fa350652503b81bdfe19ca70d224
Status...........: Exhausted

$ echo foobar | ./hashcat --quiet --potfile-disable -O -r rule 6eb3fa350652503b81bdfe19ca70d224
6eb3fa350652503b81bdfe19ca70d224:FOobar
Status...........: Cracked

$ echo foobar | ./hashcat -r rule --stdout
Foobar
$ echo foobar | ./hashcat -O -r rule --stdout
FOobar
$ echo foobar | ./hashcat -j ef --stdout
fOobar
```
To recap: The pure kernel doesn't crack the hash. The optimized kernel does crack the has AND prints the correct password. While -j still behaves the same.

Edit: It seems to be only having that issue if the first character is also the separator.
This seems to be fine: (same behavior in v 5.1.0 and current maste)
```
hops@pocl:/tmp/hashcat$ echo whatever | ./hashcat -r rule --stdout
WhatevEr
hops@pocl:/tmp/hashcat$ echo whatever | ./hashcat -O -r rule --stdout
WhatevEr
hops@pocl:/tmp/hashcat$ echo whatever | ./hashcat -j ev --stdout
WhatevEr
```

Thanks to @Chick3nman for keeping me sane while sleuthing this issue.

PS: It was initially found by @roycewilliams while cracking the CracktheCon test hsahes and submitted the wrongly reported passwords. I didn't look at it closer until 0x69BE027C97 had the exact same issue.

**Comment by jsteube** [comment]
Great Report, many thanks! Should be fixed with commit https://github.com/hashcat/hashcat/commit/ffd8ec900162156fa00022b760b339a5c66eba48

Please try on your system and close the issue if fixed.

**Comment by hops** [comment]
It is now consistent in all three rule engines.
Thinking about it a bit more and testing how john in bleeding-jumbo handles that rule maybe I was wrong with my assumption that it should mangle to `FOobar`.
This produces the same results as -j did before commit ffd8ec9
```
$ echo foobar | ./john --rules=title-sep-f --stdout --pipe
fOobar
```
I did some additional tests and found another discrepancy between hashcat and john.
If the separator occurs more than once in the input word john matches the last occurrence while hashcat does the same using -j/-k and matches all when using the pure or optimized kernel rule engine.
```
$ echo foobar | ./john --rules=title-sep-o --stdout --pipe
FooBar
$ echo foobar | ./hashcat -j eo --stdout 
FooBar
$ echo foobar | ./hashcat -r rule --stdout 
FoOBar
$ echo foobar | ./hashcat -r rule --stdout -O
FoOBar
```
In other words I don't know anymore what the correct behavior should be. Maybe @jfoug can join the discussion to give some insights into john.
IMHO I'd rather have consistency across different tools over "correctness".

**Comment by jsteube** [comment]
Thanks for testing. Since the original bug is now fixed, I'll close this issue. Any other discussion should be done in a separate issue.

**Comment by hops** [comment]
Fair enough. I'll open another issue.

**Comment by jfoug** [comment]
Here is how john works, (at this time):

$ echo foobar | ./john --stdout --pipe --rules=:ef
fOobar
1p 0:00:00:00 0.00% 4.545p/s fOobar

The words are split on separators and each separate word is cased.  The 'f' is not a word or part of a word, it is the separator.  john does not treat this as the 2 words f and oobar    Right, Wrong, or Otherwise, this is how john as of today works.  I doubt that this situation (first character being both a letter AND the separator) was looked into at all when the 'eX' rule was implemented

**Comment by jfoug** [comment]
Here is the documentation:

```
eC	title case, where C is the separation character
e?C	title case, with chars from class C as separation characters.
	NOTE, see hashcat specific 'E' rule which is generic title case for space
	only characters.  The 'e' rule extendeded more powerful version.
```

Now, it is does not document what happens IF the separator IS the first character, but I would argue (STRONGLY), that thinking that rule=ef word=foobar == FOobar is wrong.  This would really go against the logic behind this. The logic is to continually:   first-next non separator and case it and repeat.   The first letter of a word cases BECAUSE it is the first character 'after' an empty set of separators, i.e. a character leads the word, and thus is cased.  

This logic assumption could be changed to be:  repeat: first character after a block of 0 to X separators is upcased-repeat   If first letter of word is a lc letter, case it.   That is really the logic which would be needed (a special exception check at the end of that rule).  You could not do it at the start, or rule=ef word=foobar would simply become 'Foobar' which has no separators.   Also, please keep in mind, that the logic behind this is much more complex, due to john having 2 forms of the eX rule. It has eX and e?C where the ?C is a class of characters (meaning we can do e?d and then any digit is a separator, or even something as sinister as e?l meaning every lower case letter is a separator, lol.  That is what the syntax allows, it is not to say that someone 'should' use e?l

But back to original idea of this thread that foobar --> rules=:ef --> FOobar is not right, logically thinking about this.  It may be what one 'logically' sees since our brains are wired for this, but it is NOT what the program running the algorithm will produce, without changing the algorithm.

I would argue that the hashcat -j output was right to start with, UNLESS we are to document that there needs to be an exception made for ``` if(ch[0] == sep && islower(ch[0])) uc(ch[0]); ```

**Comment by hops** [comment]
@jfoug Thank you for your input! As this is now a closed issue please move the discussion over to #1970  (sorry about the confusion here).
